elastic/eui
# [EuiDataGrid] Simplify & DRY out cell/header focus

## Summary

closes https://github.com/elastic/eui/issues/5690

> Extra context: I originally spiked most of this out during the cell redesigns (#7382) probably because the repeated logic or keyboard UX slightly bothered me 😅 It turned out a larger set of changes than I wanted for a single release, so I split it out to a separate branch for later. Since it was very close to completion, I figured I'd finish it up to clean out my lingering 2023 work.

As always, please [follow along by commit](https://github.com/elastic/eui/pull/7448/commits). While it looks like more lines of code were added in this PR, those were primarily all tests (a good thing to have more of) and snapshots diffs. From actual source code, a net ~500 or so lines were removed.

### Main changes in this PR:

![datagrid_focus](https://github.com/elastic/eui/assets/549407/83955668-4b8c-47fc-aabe-d006e99ffef4)

1. I removed all logic around detecting `headerIsInteractive`. From now on, data grid headers can ALWAYS be navigated to via arrow keys, regardless of whether or not header cells have interactive content or not.
    - I believe this logic was originally added to make **EuiDataGrid** behave like the "simple" data grid example in https://www.w3.org/WAI/ARIA/apg/patterns/grid/examples/data-grids/#examples.
    - However, I don't think this is a pattern worth emulating. EuiDataGrids **don't** get used for simple data, otherwise devs would reach for EuiBasicTable instead. I strongly feel the extra code, state, and mutation observer was not worth the unnecessary UX, and frankly I also think the UX is confusing and creates inconsistencies between data grids.

2. Non-expandable cells that only contain 1 interactive child element no longer auto-focus that element, but instead render a focus trap
    - Again, I believe this logic was trying to mimic the above WCAG demo behavior, but frankly, I don't consider that behavior worth emulating. To reiterate, our data grid usages in production are vastly more complex than these simplistic demos.
    - The logic also introduced accessibility issues (see above #5690, accessibility concerns) where other non-interactive content in the cell would not be accessible to keyboard users.
    - A focus trap is overall more consistent with how other cells behave (Enter key) and more accessible to a variety of cell content.

3. Cell screen reader text now contains instructions for the `Enter` keypress behavior in addition to current column position.

For the majority of default EuiDataGrid use cases (expandable cells & column headers with actions), nothing should have changed or regressed from before.

## QA

- Go to https://eui.elastic.co/pr_7448/#/tabular-content/data-grid-cells-popovers#focus
  - [x] Confirm that pressing the `Enter` key on the non-expandable cells with interactive links/buttons inside them (both header and data cells) enters a focus trap
  - [x] Confirm that Tabbing within this focus trap only tabs to items within the cell
  - [x] Confirm the cell focus trap can be exited via the `Escape` key
- Regression testing
  - Go to https://eui.elastic.co/pr_7448/#/tabular-content/data-grid
  - [x] Confirm that tabbing in/out of the datagrid body and arrow navigation works as expected with no regressions
  - [x] Confirm that cell actions' hover/focus/popover open behavior behaves as before with no regressions

### General checklist

- Browser QA
    ~- [ ] Checked in both **light and dark** modes~
    ~- [ ] Checked in **mobile**~
- Docs site QA
    ~- [ ] Props have proper **autodocs** (using `@default` if default values are missing) and **[playground toggles](https://github.com/elastic/eui/blob/main/wiki/contributing-to-eui/documenting/playgrounds.md)**~
    ~- [ ] Checked **[Code Sandbox](https://codesandbox.io/)** works for any docs examples~
- Code quality checklist
- Release checklist
    ~- [ ] If applicable, added the **breaking change** issue label (and filled out the breaking change checklist)~
- Designer checklist
    ~- [ ] Updated the **[Figma](https://www.figma.com/community/file/964536385682658129)** library counterpart~
